src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.cs

```diff
@@ -0,0 +1,598 @@
+        }
+
+        /// <summary>
+        /// Derives the MSI package ID from the specified pack information based on the bitness of
+        /// the SDK host.
+        /// </summary>
+        /// <param name="packInfo">The pack information used to generate the package ID.</param>
+        /// <returns>The ID of the NuGet package containing the MSI corresponding to the pack.</returns>
+        private string GetMsiPackageId(PackInfo packInfo)
+        {
+            return $"{packInfo.ResolvedPackageId}.Msi.{HostArchitecture}";
+        }
+
+        /// <summary>
+        /// Downloads the NuGet package carrying the MSI and JSON manifest and extracts it locally to
+        /// a temporary folder.
+        /// </summary>
+        /// <param name="packageId">The ID of the package to download.</param>
+        /// <param name="packageVersion">The version of the package to download.</param>
+        /// <returns>The directory where the package was extracted.</returns>
+        private string DownloadAndExtractPackage(string packageId, string packageVersion)
+        {
+            string packagePath = _nugetPackageDownloader.DownloadPackageAsync(new PackageId(packageId), new NuGetVersion(packageVersion),
+                _packageSourceLocation).Result;
+
+            Log?.LogMessage($"Downloaded {packageId} ({packageVersion}) to '{packagePath}");
+
+            // Extract the contents to a random folder to avoid potential file injection/hijacking 
+            // shenanigans before moving it to the final cache directory.
+            string extractionDirectory = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
+            Directory.CreateDirectory(extractionDirectory);
+            Log?.LogMessage($"Extracting '{packageId}' to '{extractionDirectory}'");
+            _ = _nugetPackageDownloader.ExtractPackageAsync(packagePath, new DirectoryPath(extractionDirectory)).Result;
+
+            return extractionDirectory;
+        }
+
+        /// <summary>
+        /// Gets a set of all the installed SDK feature bands.
+        /// </summary>
+        /// <returns>A List of all the installed SDK feature bands.</returns>
+        private IEnumerable<SdkFeatureBand> GetInstalledFeatureBands()
+        {
+            HashSet<SdkFeatureBand> installedFeatureBands = new();
+            foreach (string sdkVersion in GetInstalledSdkVersions())
+            {
+                try
+                {
+                    installedFeatureBands.Add(new SdkFeatureBand(sdkVersion));
+                }
+                catch (Exception e)
+                {
+                    Log?.LogMessage($"Failed to map SDK version {sdkVersion} to a feature band. ({e.Message})");
+                }
+            }
+
+            return installedFeatureBands;
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        protected virtual void Dispose(bool disposing)
```

Comment:
What is the disposing param for? It looks like it's never used and the name is unclear.

Reply:
the overload with the bool is required for non-sealed classes to handle managed/unmanaged resources. I added this long ago when I was trying to figure out when a command was completed. I figured once the command itself was being disposed it might be a good indication to shutdown the server.

I've removed this